### ISSUE TITLE
인공지능 - 머신러닝 -딥러닝

### ISSUE THREAD

**Issue by skarltjr** [skarltjr/Memory_Write_Record#22]

**Comment by skarltjr** [comment]
참고 :
딥러닝 히스토리 https://machinelearningknowledge.ai/brief-history-of-deep-learning/#Neocognitron_8211_First_CNN_Architecture
딥러닝 비교 : https://dbrang.tistory.com/1537

**Comment by skarltjr** [comment]
### 딥러닝 프레임워크 ( Torch / TensorFlow)
1. 프레임워크 
- TensorFlow : 구글팀에서 개발했으며 2015년도 오픈소스로 공개되었다.
- Pytorch :  16년 10 월 공개. 페이스북 AI 리서치 랩 (FAIR). torch에 기반하여 제작
2. 플랫폼, 인터페이스 비교

|프레임워크|Platform|written in|supported interface
|------|---|---|---|
|TensorFlow|Linux, Mac OS| C++, Python|C++, Python|
|Torch|Linux, Android, Mac OS,iOS, Windows|C, Lua| C, C++, Lua|

3. 선행학습 모델, RNN, CNN 지원여부 비교

|프레임워크|has pretrained model|RNN|CNN
|------|---|---|---|
|TensorFlow|yes| yes|yes|
|Torch|yes|yes| yes|


<img width="1194" alt="Screen Shot 2021-07-29 at 9 16 38 AM" src="https://user-images.githubusercontent.com/62214428/127412358-e0779932-1726-48f4-bacb-3c043b5ddfa7.png">

**Comment by skarltjr** [comment]
## 1. 인공지능 / 머신러닝 / 딥러닝
### 1-1 인공지능
- 1956년, 미국 다트머스 대학에 있던 존 매카시 교수가 개최한 다트머스 회의를 통해 처음으로 인공지능(AI)라는 용어가 사용되기 시작
    -  당시 인공지능 연구의 핵심은 추론과 탐색
### 1-2 머신러닝
- 1959년 아서 사무엘은 머신 러닝을 "컴퓨터가 명시적으로 프로그램되지 않고도 학습할 수 있도록 하는 연구 분야"라고 정의.

- 1998년 카네기 멜론 대학의 톰 미첼은 "만약 어떤 작업 T에서 경험 E를 통해 성능 측정 방법인 P로 측정했을 때 성능이 향상된다면 이런 컴퓨터 프로그램은 학습을 한다고 말한다" 라고 프로그램의 학습을 정의

- 머신러닝은 지도학습, 비지도학습과 강화학습의 세 가지 종류가 있다.
     - 지도학습 : 입력값과 함께 결과값을 같이 주고 학습을 시키는 방법으로, 분류 / 회귀 등 여러가지 방법에 사용된다. 
         - 지도학습의 구체적인 학습모델은 신경망, SVM(Support Vector Machine), 결정트리(Decision Tree), 회귀모델 등이 있다.

     - 비지도학습 : 지도 학습과는 달리 정답 라벨이 없는 데이터를 비슷한 특징끼리 군집화 하여 새로운 데이터에 대한 결과를 예측하는 방법을 비지도학습 이라고 한다.
         - 비지도학습의 대표적인 종류는 클러스터링(Clustering)이 있다. 

     - 강화학습 : 행동 심리학에서 나온 이론으로 분류할 수 있는 데이터가 존재하는 것도 아니고 데이터가 있어도 정답이 따로 정해져 있지 않으며 자신이 한 행동에 대해 보상(reward)를 받으며 학습하는 것을 말한다.
          - 강화학습에 딥러닝을 성공적으로 적용한 대표적 알고리즘으로는 DQN과 A3C가 있다. 두 알고리즘 모두 딥마인드에서 발표했으며 다른 강화학습 알고리즘의 베이스라인이 되었다.



### 1-3 딥러닝
- 딥 러닝(deep learning)은 머신 러닝의 한 분야로서 다량의데이터나 복잡한 자료들 속에서 핵심적인 내용 또는 기능을 요약하는 작업을 하는 알고리즘의 집합이다.

## 2. 딥러닝의 시초
- 1943년 최초로 제안된 인공신경망(Artificial Neural Networks)은 인간의 뇌 신경세포 구조를 수학적으로 모델링하는 개념
- ![화면 캡처 2021-07-28 234513](https://user-images.githubusercontent.com/62214428/127343122-82eca14d-e707-40ed-aa1a-2f311bbb1c46.png)
  1) 인간의 뉴런은 Dendrite(수상돌기)를 통해 여러 개의 시그널을 받는다
  2) Dendrite가 받은 여러 개의 시그널이 복합적으로 계산되어 Cell body에 전달되면 한 가지의 시그널이 만들어지고
  3) 만들어진 한 개의 시그널이 Axon(축색돌기)을 통해 다른 뉴런으로 전달
  - ![화면 캡처 2021-07-28 234701](https://user-images.githubusercontent.com/62214428/127343436-e344f545-fab2-42a2-a316-b6e1703d3816.png)
  4) Dendrite와 Axon terminal의 결합점인 시냅스(Synapse)에서 일정치 이상의 입력신호를 받으면 출력신호가 나타난다.
#### 이것이 우리의 뇌가 신호를 받아서 반응하는 원리

## 2-1. perceptron
![화면 캡처 2021-07-28 234850](https://user-images.githubusercontent.com/62214428/127343732-3424d718-9158-4390-87b7-8f697b76be70.png)
- 최초의 인공신경망 모델인 퍼셉트론은 위처럼 인간의 뉴런과 비슷한 구조
- 퍼셉트론 모델이 x라는 신호(input)를 받으면 각각의 x는 각각의 가중치(w)와 곱해지고 편향치(b)와 더해진다.
- 모든 인풋 결과가 더해진 값을 활성화 함수에 인자로 대입하면 기대치 여부에 따라 다른 신호(output)가 리턴
![화면 캡처 2021-07-28 235025](https://user-images.githubusercontent.com/62214428/127344024-ea0c63cf-5f43-445a-958e-250c56dbebe4.png)

## 2-2 neural network의 종류
- ANN
    - 머신러닝의 한 분야인 딥러닝은 인공신경망(Artificial Neural Network)를 기초로 하고있다. 
    - ANN은 사람의 신경망 원리와 구조를 모방하여 만든 기계학습 알고리즘으로 인간의 뇌에서 뉴런들이 어떤 신호, 자극(input) 등을 받고, 그 자극이 어떠한 임계값(weight)을 넘어서면 결과 신호(output)를 전달하는 과정에서 착안한 것
    - 문제점 : 
        - 학습과정에서 파라미터의 최적값을 찾기 어렵다.
        - 은닉층이 많으면 정확도가 증가하지만 연산량이 기하 급수적으로 증가하고 매우 느려진다.

- DNN
    - 1965년 Alexey Grigoryevich Ivakhnenko가  처음으로 다계층 인공신경망 알고리즘을 만들었다.
    - 심층 신경망(Deep Neural Network, DNN)은 입력 계층(input layer)과 출력 계층(output layer) 사이에 복수개의 은닉 계층(hidden layer)들로 이뤄진 인공신경망이다.
    - ANN의 문제점이 해결되며 모델 내 은닉층을 많이 늘려서 학습의 결과를 향상시키는 방법
    - 컴퓨터가 스스로 분류 레이블을 만들어내고 공간을 왜곡하고 데이터를 구분짓는 과정을 반복하여 최적의 구분선을 도출
    - <img width="543" alt="Screen Shot 2021-07-28 at 4 04 26 PM" src="https://user-images.githubusercontent.com/62214428/127279313-504ceb61-3833-4c1d-b43c-eb671353d64e.png">


-  CNN
    - 1980년 Kunihiko Fukushima가 초기 cnn인 Neocognitron을 만들었다.
    - 합성곱 신경망(Convolutional Neural Network, CNN)은 최소한의 전처리(preprocess)를 사용하도록 설계된 다계층 퍼셉트론의 한 종류이다
        - 참고 : 퍼셉트론(Perceptron)은 프랑크 로젠블라트(Frank Rosenblatt)가 1957년에 제안한 초기 형태의 인공 신경망으로 다수의 입력으로부터 하나의 결과를 내보내는 알고리즘 
     - 기존의 방식은 데이터에서 지식을 추출해 학습이 이루어졌지만, CNN은 데이터의 특징을 추출하여 특징들의 패턴을 파악하는 구조. 이 CNN 알고리즘은 Convolution과정과 Pooling과정을 통해 진행
          - Convolution : 데이터의 특징을 추출하는 과정으로 데이터에 각 성분의 인접 성분들을 조사해 특징을 파악하고 파악한 특징을 한장으로 도출시키는 과정. 여기서 도출된 장을 Convolution Layer. 이 과정은 하나의 압축 과정이며 파라미터의갯수를 효과적으로 줄여주는 역할.
          - <img width="665" alt="Screen Shot 2021-07-28 at 4 07 18 PM" src="https://user-images.githubusercontent.com/62214428/127279275-529e6f40-800b-4eaf-86cf-f69c852e1536.png">

          - Pooling : Convolution 과정을 거친 레이어의 사이즈를 줄여주는 과정. 단순히 데이터의 사이즈를 줄여주고, 노이즈를 상쇄시키고 미세한 부분에서 일관적인 특징을 제공.


          


 - RNN
    - 1982년 John Hopfield가 rnn의 원조인 Hopfield Network를 개발.
    - 순환 신경망(Recurrent Neural Network, RNN)은 인공신경망을 구성하는 유닛 사이의 연결이 Directed cycle을 구성하는 신경망을 말한다
    - RNN 알고리즘은 반복적이고 순차적인 데이터(Sequential data)학습에 특화된 인공신경망의 한 종류로써 내부에 순환구조가 들어있다는 특징을 가지고 있다
    - 순환구조를 이용하여 과거의 학습을 Weight를 통해 현재 학습에 반영. 
        - 현재의 학습과 과거의 학습의 연결을 가능하게 하고 시간에 종속된다는 특징. 
        - 여러 단계에서 매개 변수 공유 → 훈련 매개 변수 감소 및 계산 비용 감소
    - 음성 웨이브폼을 파악하거나, 텍스트의 앞 뒤 성분을 파악할 때 주로 사용
    - <img width="579" alt="Screen Shot 2021-07-28 at 4 08 33 PM" src="https://user-images.githubusercontent.com/62214428/127279435-4d7558f6-b91d-4da1-aef7-a1ee70c118bc.png">

![화면 캡처 2021-07-28 214418](https://user-images.githubusercontent.com/62214428/127324159-e06e3a69-9e8b-4b76-86e3-7c26c71e00a5.png)

## 3 neural network history
#### 1957 - Setting the foundation for deep neural networks
- 언제 : 1957
- 누가 : Frank Rosenblatt
- “The Perceptron: A Perceiving and Recognizing Automaton” 제안
- 퍼셉트론(Perceptron)은 프랑크 로젠블라트(Frank Rosenblatt)가 1957년에 제안한 초기 형태의 인공 신경망으로 다수의 입력으로부터 하나의 결과를 내보내는 알고리즘
- 그는 뇌의 지각 과정과 유사한 방식으로 시각, 전기, 음색 정보 패턴의 유사성과 동일성을 인식하는 방법을 학습하는 전기, 전자 시스템을 구축할 수 있다고 말했다. 
- 이는 소프트웨어보단 하드웨어적 알고리즘에 가깝지만 deep neural networks (DNN)의 기반으로 자리잡았다. 

#### 1959 - ANN / Discovery of simple cells and complex cells
- 언제 : 1959
- 누가 : David H. Hubel and Torsten Wiesel
- neurophysiologists and Nobel Laureates David H. Hubel and Torsten Wiesel discovered two types of cells in the primary visual cortex: simple cells and complex cells.
- artificial neural networks (ANNs) : ANN은 이런 생물학적 관점으로부터 시작되었고 이 분야에 매우 큰 영향

#### 1965 - DNN의 시초인  Group Method of Data Handling (GMDH) / The first working deep learning networks
- 언제 : 1965
- 누가 : Ivakhnenko 
-  Group Method of Data Handling (GMDH)을 개발 
-  Group Method of Data Handling (GMDH) : 컴퓨터 기반 다중 파라미터 datasets의 수학적 모델링을 위한 유도 알고리즘 집합체로 완전한 자동 구조와 모델의 매개 변수 최적화가 특징이다.
- GMDH를 이용하여 Ivakhnenko는 1971년 8-layer deep network를 개발했고 이를 통해 Alpha라고 불리는 computer identification system의 learning process를 성공적으로 입증했다.

#### 1980 - An ANN learns how to recognize visual patterns
- 언제 : 1980
- 누가 : Kunihiko Fukushima
- 시각적인 패턴 인지를 학습하는 artificial neural network인  Neocognitron을 개발 
- Kunihiko Fukushima는 앞의 ANN의 기반을 마련한 Hubel and Wiesel로 부터 영향을 받아 동물로부터 발견된 시각 피질 조직(visual cortex organization)을 바탕으로 CNN을 개발을 이끌었다.
- 합성곱 신경망(Convolutional Neural Network, CNN)은 최소한의 전처리(preprocess)를 사용하도록 설계된 다계층 퍼셉트론의 한 종류이다

#### 1982 - The creation of the Hopfield Networks / RNN
- 언제 :1982
- 누가 : John Hopfield
- Hopfield Networks are a recurrent neural network( RNN ) that serve as a content-addressable memory system
- 순환 신경망(Recurrent Neural Network, RNN)은 인공신경망을 구성하는 유닛 사이의 연결이 Directed cycle을 구성하는 신경망을 말한다
- RNN 알고리즘은 반복적이고 순차적인 데이터(Sequential data)학습에 특화된 인공신경망의 한 종류로써 내부에 순환구조가 들어있다는 특징을 가지고 있다
- 순환구조를 이용하여 과거의 학습을 Weight를 통해 현재 학습에 반영.
    - 현재의 학습과 과거의 학습의 연결을 가능하게 하고 시간에 종속된다는 특징.
    - 여러 단계에서 매개 변수 공유 → 훈련 매개 변수 감소 및 계산 비용 감소


참고 :
딥러닝 히스토리 
https://machinelearningknowledge.ai/brief-history-of-deep-learning/#Neocognitron_8211_First_CNN_Architecture
https://www.import.io/post/history-of-deep-learning/

딥러닝 비교 : https://dbrang.tistory.com/1537
